components/builder-originsrv/src/migrations/origin_packages.rs

```diff
@@ -267,5 +317,32 @@ pub fn migrate(migrator: &mut Migrator) -> Result<()> {
                         RETURN;
                     END
                     $$ LANGUAGE plpgsql STABLE"#)?;
+    migrator.migrate("originsrv",
+                     r#"CREATE OR REPLACE FUNCTION get_origin_packages_for_version_v1 (
+                            op_ident text,
+                            op_limit bigint,
+                            op_offset bigint
+                        ) RETURNS TABLE(total_count bigint, ident text, channels text[]) AS
+                        $$
+                            BEGIN
+                                RETURN QUERY
+                                   SELECT COUNT(*) OVER ()
+                                       AS total_count,
+                                          origin_packages.ident,
+                                          (SELECT ARRAY(SELECT name
+                                                          FROM origin_channels
+                                                         WHERE id IN (SELECT channel_id
+                                                                        FROM origin_channel_packages
+                                                                       WHERE package_id = origin_packages.id
+                                                                       ORDER BY created_at DESC)))
+                                     FROM origin_packages
+                                    WHERE origin_packages.ident LIKE (op_ident  || '%')
+                                 ORDER BY ident DESC
+                                    LIMIT op_limit
+                                   OFFSET op_offset;
+                                RETURN;
```

Comment:
I think this query will perform better when formulated using [Common Table Expressions](https://www.postgresql.org/docs/current/static/queries-with.html). As currently written, we'll end up performing the subquery for every identifier that matches the query string, even if they will ultimately end up falling outside of our `LIMIT` / `OFFSET` window. Depending on the query string (e.g. simply "core"), that could amount to a lot of unnecessary work.

With a CTE, which you can roughly treat as a temporary table that exists for the duration of a single query, you could quickly narrow down the data to just what satisfies the query and what falls within the window, and then use that to perform your subqueries against.

Happy to give some more pointers on this if you like.